### PR TITLE
Add Caffeine cache support

### PR DESCRIPTION
**Pull request by eddumelendez**
See gh-4899

### CODE REVIEW

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CaffeineCacheConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * Caffeine cache configuration.
+ *
+ * @author Eddú Meléndez
+ * @since 1.4.0
+ */
+@Configuration
+@ConditionalOnClass({ Caffeine.class, CaffeineCacheManager.class })
+@ConditionalOnMissingBean(org.springframework.cache.CacheManager.class)
+@Conditional({ CacheCondition.class,
+		EhCacheCacheConfiguration.ConfigAvailableCondition.class })
```
Comment (shakuzen) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CaffeineCacheConfiguration.java]: `EhCacheCacheConfiguration`? Looking at some of the other Configuration classes, it seems like maybe this should be `@Conditional(CacheCondition.class)` unless I'm missing something.

**Comment by ben-manes** [comment]
Thanks for this. What is the idiomatic way to specify the configuration (size, expiration, etc) in Spring? I'm a little outdated on current idioms, having not used Spring extensively since the 2.x days.

I think @snicoll has wisdom here. I had [thought](https://jira.spring.io/browse/SPR-13690) a Guava-style [CacheBuilderSpec](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/cache/CacheBuilderSpec.html) might be desired, but that appears not to be the case. Let me know if there's anything I can do to help out.

**Comment by eddumelendez** [comment]
Hi @ben-manes you can create a bean like [this](https://github.com/eddumelendez/spring-boot/blob/gh-4899/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfigurationTests.java#L799-L802) and the auto-configuration will take this.

**Comment by ben-manes** [comment]
Oh cool! I assume that can be scoped for multiple instances/configurations using named beans and `@Qualifier`? My last use was with xml wiring and writing custom namespaces parsers to simplify configuration, before moving into the guice/dagger world at G.

**Comment by ben-manes** [comment]
Found a [gist](https://gist.github.com/snicoll/7bf72686c98a7088e8ea) by @snicoll, so all of this makes sense to me now. :+1:

**Comment by snicoll** [comment]
@ben-manes I am not against a `CacheBuilderSpec` at all. If you feel that Caffeine lacks of configuration options, feel free to add whatever you think is best. Having said that, I think that the `CacheBuilder` @eddumelendez referenced is really nice.

**Comment by ben-manes** [comment]
@eddumelendez looks perfect! I didn't mean to imply anything except my own naivety :-)

**Comment by ben-manes** [comment]
Some Spring users requested [CaffeineSpec](http://static.javadoc.io/com.github.ben-manes.caffeine/caffeine/2.1.0/com/github/benmanes/caffeine/cache/CaffeineSpec.html) be provided to compliment @eddumelendez's excellent addition for externalizing their configuration settings. I released version [2.1.0](https://github.com/ben-manes/caffeine/releases/tag/v2.1.0) with that improvement.

**Comment by snicoll** [comment]
:+1: Eddu,I can take care of that change when I'll merge your PR or you can push force here if you want to do it.

**Comment by eddumelendez** [comment]
@ben-manes @snicoll improvement has been added in the PR :smiley:

**Comment by eddumelendez** [comment]
@snicoll PR updated after review your comments in https://github.com/eddumelendez/spring-boot/commit/36939ff30d34ba0ddb256ab83c1ac7cfeab4b854

**Comment by snicoll** [comment]
Thanks again for another great PR Eddu. This is merged with some polish (see f04b517). I've also updated to the latest release.

**Pull request closed without merging**